### PR TITLE
5565 allow edit of file position in directory structure

### PR DESCRIPTION
**Pull request by sekmiller**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5565 Edit Position of files in directory structure

## Pull Request Checklist

- [x] Unit [tests][x]none
- [x] Integration [tests][x]: completed
- [x] Deployment requirements,: None
- [ ] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][x] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: doc/sphinx-guides/source/user/dataset-management.rst [hunk 1]
```diff
@@ -180,11 +180,13 @@ Metadata found in the header section of `Flexible Image Transport System (FITS)
 Compressed Files
 ----------------
 
-Compressed files in zip format are unpacked automatically. If it fails to unpack, for whatever reason, it will upload as is. If the number of files inside are more than a set limit (1,000 by default, configurable by the Administrator), you will get an error message and the zip file will uploads as is.
+Compressed files in .zip format are unpacked automatically. If a .zip file fails to unpack for whatever reason, it will upload as is. If the number of files inside are more than a set limit (1,000 by default, configurable by the Administrator), you will get an error message and the .zip file will upload as is.
+
+If the uploaded .zip file contains a folder structure, Dataverse will keep track of this structure. A file's location within this folder structure is displayed in the file metadata as the File Path. When you download the contents of the dataset, this folder structure will be preserved and files will appear in their original locations.
```
Comment (pdurbin) [inline doc/sphinx-guides/source/user/dataset-management.rst]: Do I get the exact same folder structure as I originally uploaded? Here's an example of how we changed directory names and sometimes file names in the DVN 3 days: https://github.com/IQSS/dataverse-client-r/issues/18#issuecomment-327268497

## The zip file I uploaded to DVN 3

```
murphy:zip pdurbin$ zip -r data.zip folder1 folder2 -x "*.DS_Store"
  adding: folder1/ (stored 0%)
  adding: folder1/sub1/ (stored 0%)
  adding: folder1/sub1/file.txt (stored 0%)
  adding: folder1/sub2/ (stored 0%)
  adding: folder1/sub2/file.txt (stored 0%)
  adding: folder2/ (stored 0%)
  adding: folder2/sub1/ (stored 0%)
  adding: folder2/sub1/file.txt (stored 0%)
  adding: folder2/sub2/ (stored 0%)
  adding: folder2/sub2/file.txt (stored 0%)
murphy:zip pdurbin$ 
```

## The zip file I downloaded from DVN 3

```
murphy:unzip pdurbin$ unzip study_10000.zip 
Archive:  study_10000.zip
  inflating: folder1-sub1/file.txt   
  inflating: folder1-sub2/file_1.txt  
  inflating: folder2-sub1/file_2.txt  
  inflating: folder2-sub2/file_3.txt  
  inflating: MANIFEST.TXT            
murphy:unzip pdurbin$ 
```

File: src/main/java/edu/harvard/iq/dataverse/EditDatafilesPage.java [hunk 2]
```diff
@@ -2760,6 +2784,32 @@ public void handleDescriptionChange(final AjaxBehaviorEvent event) {
...
+        
+        if (!constraintViolations.isEmpty()) {
+             //JsfHelper.addFlashMessage(getBundleString("dataset.message.validationError"));
+            logger.fine("Constraint violation detected on SAVE: "+constraintViolations.toString());
+             JH.addMessage(FacesMessage.SEVERITY_ERROR, getBundleString("dataset.message.validationError"));
+             
+            //FacesContext.getCurrentInstance().addMessage(null, new FacesMessage(FacesMessage.SEVERITY_ERROR, "Validation Error", "See below for details."));
+
+        }
+               */
```
[the remark this answers is not in the document]
Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/EditDatafilesPage.java]: Sure. thanks.

File: doc/sphinx-guides/source/user/dataset-management.rst (hunk 1 above)
Comment (pdurbin) [reply doc/sphinx-guides/source/user/dataset-management.rst]: I just tested this on demo (thanks for the suggestion at standup) and here's how it works as of Dataverse 4.11.

## The zip file I uploaded to Dataverse 4.11

```
murphy:uploaded pdurbin$ unzip data.zip 
Archive:  data.zip
   creating: folder1/
   creating: folder1/sub1/
 extracting: folder1/sub1/file.txt   
   creating: folder1/sub2/
 extracting: folder1/sub2/file.txt   
   creating: folder2/
   creating: folder2/sub1/
 extracting: folder2/sub1/file.txt   
   creating: folder2/sub2/
 extracting: folder2/sub2/file.txt   
murphy:uploaded pdurbin$ 
```

## The zip file I downloaded from Dataverse 4.11

```
murphy:downloaded pdurbin$ unzip dataverse_files.zip 
Archive:  dataverse_files.zip
   creating: folder1/sub1/
  inflating: folder1/sub1/file.txt   
   creating: folder1/sub2/
  inflating: folder1/sub2/file.txt   
   creating: folder2/sub1/
  inflating: folder2/sub1/file.txt   
   creating: folder2/sub2/
  inflating: folder2/sub2/file.txt   
  inflating: MANIFEST.TXT            
murphy:downloaded pdurbin$ 
```

This is looking good to me. Dataverse adds the MANIFEST.TXT file, but this is a feature. Here is the content of that file:

```
murphy:downloaded pdurbin$ cat MANIFEST.TXT 
folder1/sub1/file.txt (text/plain) 16 bytes.
folder1/sub2/file.txt (text/plain) 16 bytes.
folder2/sub1/file.txt (text/plain) 13 bytes.
folder2/sub2/file.txt (text/plain) 13 bytes.
murphy:downloaded pdurbin$ 
```

File: src/main/java/edu/harvard/iq/dataverse/util/StringUtil.java [hunk 3]
```diff
@@ -180,6 +180,19 @@ public static String decrypt(String value, String password ) {
...
     
+    public static String sanitizeFileDirectory(String value){
+        
+        while (value.startsWith("\\") || value.startsWith("/") || value.startsWith("-") || value.startsWith(".")){
+            value = value.substring(1);
+        }
+        while (value.endsWith("\\") || value.endsWith("/") || value.endsWith("-") || value.endsWith(".")){
+            value = value.substring(0, value.length() - 1);
+        }
+
```
Comment (landreev) [inline src/main/java/edu/harvard/iq/dataverse/util/StringUtil.java]: @sekmiller This method correctly strips the leading and trailing slashes (and also "." and "-"); 
But I thought the plan was also to replace any multiple slashes between nested folders with a single slash. 
For example, as implemented now, I can enter "folder1///folder2", and it gets saved and displayed like this, with the 3 slashes. 
I thought we wanted to sanitize it as "folder1/folder2".

Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/util/StringUtil.java]: Added a remove of multiple inner slashes to the sanitizer as suggested. Thanks @landreev

**Pull request merged**